cmd/gateway/b2/gateway-b2.go

```diff
@@ -595,14 +604,10 @@ func (l *b2Objects) DeleteObject(ctx context.Context, bucket string, object stri
 	if err != nil {
 		return err
 	}
-	reader, err := bkt.DownloadFileByName(l.ctx, object, 0, 1)
-	if err != nil {
-		logger.LogIf(ctx, err)
-		return b2ToObjectError(err, bucket, object)
-	}
-	io.Copy(ioutil.Discard, reader)
-	reader.Close()
-	err = bkt.File(reader.ID, object).DeleteFileVersion(l.ctx)
+
+	// If we hide the file we'll conform to B2's versioning policy, it also
+	// saves an additional call to check if the file exists first
+	_, err = bkt.HideFile(l.ctx, object)
```

Comment:
This will never free space stored in B2 backend, also I am not sure why you are modifying this in the first place, can you explain ?

Reply:
The way that `DeleteFileVersion` functions means that if a file has been overwritten multiple times it will only delete the most recent version of the file, promote the next-most-recent to the be primary and the file will appear to remain in the bucket - even if the lifecycle is set to "keep only the last version of the file".

The versions are only cleaned up every 24 hours at a maximum so `DeleteFileVersion` may only function as expected every 24 hours.

I initially worked around this on a previous commit by iterating through all the file versions and deleting them all, thus simulating the expected behaviour.

This seemed a little aggressive though and would break the user's expectation if they'd actively configured a custom lifecycle on a bucket so I thought this approach would be better.

B2 cleans up the storage space on the basis of the lifecycle settings of the bucket - if a bucket is set to "keep only the last version of a file" hiding the file causes it to be deleted after 24 hours (the minimum) but removes it from being visible via the API.

Any other custom lifecycle settings would cause the hidden file to be cleaned up on the basis of those settings.

B2 seems thoroughly weird in the way their versioning and cleanup work, I'm happy to revert to the hard "iterate through all versions" method if you think it will match the expectations of MinIO better though.

**Reference**:

https://www.backblaze.com/b2/docs/file_versions.html
https://www.backblaze.com/b2/docs/lifecycle_rules.html

Reply:
The alternative function that forcefully deletes all versions looks like:

```Go
func (l *b2Objects) DeleteObject(ctx context.Context, bucket string, object string) error {
	bkt, err := l.Bucket(ctx, bucket)
	if err != nil {
		return err
	}

	f, _, err := bkt.ListFileNames(l.ctx, 1, object, "", "")
	if err != nil {
		logger.LogIf(ctx, err)
		return b2ToObjectError(err, bucket, object)
	}

	// B2's list will return the next item in the bucket if the object doesn't
	// exist so we need to perform a name check too
	if len(f) != 1 || (len(f) == 1 && f[0].Name != object) {
		return minio.ObjectNotFound{
			Bucket: bucket,
			Object: object,
		}
	}

	for {
		// B2's versions stick around for 24 hours (even in unversioned mode), by iterating
		// through all the versions on delete we better simulate S3 behaviour
		v, _, _, err := bkt.ListFileVersions(l.ctx, 1, object, "", "", "")
		if err != nil {
			logger.LogIf(ctx, err)
			return b2ToObjectError(err, bucket, object)
		}

		if len(v) != 1 || (len(v) == 1 && v[0].Name != object) {
			return nil
		}

		err = bkt.File(v[0].ID, object).DeleteFileVersion(l.ctx)
		if err != nil {
			logger.LogIf(ctx, err)
			return b2ToObjectError(err, bucket, object)
		}
	}
}
```